### ISSUE TITLE
Default combination of fnTimeout and Idletimeout causing http status code 504 on parallel executions

### ISSUE THREAD

**Issue by hd40910** [fission/fission#1673]
@vishal-biyani 

In one of our tests we are trying to invoke the same poolmamanger function in parallel and it fails after few invocations.
**The steps we followed are :**
- Create a python function, the code has a random sleep between 10 to 30 seconds
- Create a spec for this function and deploy it will default  fnTimeout and Idletimeout
- Invoke this function via http route using 10 parallel threads for 10 times
- After few invocations we receive a 504 with error : function not responses before the timeout

Question is if there is a better way to play around it, we expect the same function to be executed multiple times from multiple sources at the same time.
We also tried increasing the fntimeout and Idletimeout= fntimeout * 2 to a higher number to control the behavior.
This fixes the issue initially for few more invocations but fails after some time once there are multiple request in queue
Any suggestions to control the behavior ?

> Router Error log :
```
2020-07-22T14:29:14.478Z        ERROR   triggerset.http_trigger_set.test-fun-1  router/functionHandler.go:646   **function not responses before the timeout**        {"function": {"metadata":{"name":"test-fun-1","namespace":"default","selfLink":"/apis/fission.io/v1/namespaces/default/functions/test-fun-1","uid":"8acaf4b9-71f6-4f5b-859f-49b581c8e881","resourceVersion":"199363","generation":1,"creationTimestamp":"2020-07-16T11:01:26Z","annotations":{"fission-name":"test-fun-1","fission-uid":"default-atomiq-py-database-test-fun-1"}},"spec":{"environment":{"namespace":"fission-function","name":"atomiq-py-database"},"package":{"packageref":{"namespace":"default","name":"test-fun-1","resourceversion":"199352"}},"secrets":null,"configmaps":[{"namespace":"default","name":"account-configmap"}],"resources":{},"InvokeStrategy":{"ExecutionStrategy":{"ExecutorType":"poolmgr","MinScale":0,"MaxScale":0,"TargetCPUPercent":0,"**SpecializationTimeout":120**},"StrategyType":"execution"},**"functionTimeout":60,"idletimeout":120**}}, "request_header": {"Accept":["*/*"],"Accept-Encoding":["gzip, deflate"],"Content-Length":["2196"],"Content-Type":["application/json"],"User-Agent":["python-requests/2.24.0"],"X-Fission-Full-Url":["/default/atomiq_py_database/test_fun_1"],"X-Fission-Function-Name":["test-fun-1"],"X-Fission-Function-Namespace":["default"],"X-Fission-Function-Resourceversion":["199363"],"X-Fission-Function-Uid":["8acaf4b9-71f6-4f5b-859f-49b581c8e881"],"X-Forwarded-For":["10.78.139.116, 10.244.0.16"],"X-Forwarded-Host":["ilgss0508:30001"],"X-Forwarded-Port":["443"],"X-Forwarded-Proto":["https"],"X-Real-Ip":["10.78.139.116"],"X-Request-Id":["8203489aaf4703021c1b959cfcc626d4"],"X-Scheme":["https"]}}
```
> Spec:
```
Name:         test-fun-1
Namespace:    default
Labels:       <none>
Annotations:  fission-name: test-fun-1
              fission-uid: default-test-fun-1
API Version:  fission.io/v1
Kind:         Function
Metadata:
  Creation Timestamp:  2020-07-16T10:40:01Z
  Generation:          1
  Resource Version:    193580
  Self Link:           /apis/fission.io/v1/namespaces/default/functions/test-fun-1
  UID:                 7fbe4fdf-da61-44ca-a1df-bd5e983bc74a
Spec:
  Invoke Strategy:
    Execution Strategy:
      Executor Type:           poolmgr
      Max Scale:               0
      Min Scale:               0
      Specialization Timeout:  120
      Target CPU Percent:      0
    Strategy Type:             execution
  Configmaps:
    Name:       account-configmap
    Namespace:  default
  Environment:
    Name:            atomiq-py-ssh
    Namespace:       fission-function
  Function Timeout:  60
  Idletimeout:       120
  Package:
    Packageref:
      Name:             test-fun-1
      Namespace:        default
      Resourceversion:  193574
  Resources:
  Secrets:  <nil>
Events:     <none>
```

**Comment by vishal-biyani** [comment]
Hey @hd40910 This is an error due to `SpecializationTimeout` - and there is a configuration in function spec. Can you please try increasing that to something like 300s and LMK the results please?

```
spec:
  InvokeStrategy:
    ExecutionStrategy:
      ExecutorType: poolmgr
      MaxScale: 0
      MinScale: 0
      SpecializationTimeout: 120**
      TargetCPUPercent: 0
    StrategyType: execution
  configmaps: null
```

**Comment by hd40910** [comment]
Thanks @vishal-biyani 

SpecializationTimeout did not help with this .

After trying the suggestion and playing with the timeouts, I have more details to add regarding the issue. Am pin pointing to the usage of  idletimeout

There are 3 things in picture:
1.	Fntimeout – Time to wait for function to respond
2.	Idletimeout – Time pod will stay ideal if no new request is received
3.	Graceperiod – Time to wait before the Terminating pod is killed

**Specs:**
Fntimeout – 3600 – A very big number so that we will not get a timeout
Ideltimeout – 60 seconds – After a pod gets spawned to ‘Running’ and it does not receive a new request within 60 seconds. It will be move to Terminating state.

**Steps :** The function is a dummy function with static 30 sec sleep in its code.

[Time: 0 sec] Start 20 threads in parallel and invoke the same function via http route  – The pod status is “Running”. This pod is going to process all the 20 requests sequentially
[Time: After 10 sec] All the threads are in waiting status . They are waiting for thread 1 to complete. – Please Note: The pod status is still “Running”
[Time: After 30 sec] Thread 1 completes as function has responded. Function starts working on Thread2  – Please Note: The pod status is still “Running”
[Time: After 60 sec] Thread 2 completes as function has responded. Function starts working on Thread3  – Please Note: The pod status is now  “Terminating”  as no new request have occurred (But it’s still processing old requests)
[Time: After 90 sec] Thread 3 completes as function has responded. Function starts working on Thread4  – Please Note: The pod is in  “Terminating” (But it’s still processing old requests)
.
.
.
[Time: After 380 sec] Start another thread and invoke the function via http route. Please Note: This request goes to a new pod. The new pod status is “Running”. The old pod is still processing old requests.
.
.
[Time: After 390 sec] Thread 13 completes as function has responded. Function starts working on Thread 14  – Please Note: The old pod is in  “Terminating” (But it’s still processing old requests)
[Time: After 420 sec] Grace Period reached and the old “Terminating” pods gets deleted.
[Time: After 420 sec] **Thread 15 – 20 gets marked as failed as soon as the old pod gets deleted.**
[Time: After 430 sec] Start another thread and invoke a function via http route. Please Note: This request goes to the new pod as its still in Running status.


**Looks like the usage of these timeout is different than expected. Can you please help clarify the same ?**

**Comment by vishal-biyani** [comment]
@hd40910 This is indeed a bug in Fission and needs to be fixed. The basic problem is right now when the same function is called in the pool manager, only one specialization is happening. I will prioritize this to be of higher priority in upcoming releases.

**Comment by hd40910** [comment]
Thanks @vishal-biyani.

**Comment by vishal-biyani** [comment]
Closing this as this is now released in 1.11.0
